MaryConfigurationException: Cannot start MARY server

`MaryConfig.getMainConfig()` returns null which leads to the following stack trace:

```
marytts.exceptions.MaryConfigurationException: Cannot start MARY server
	at marytts.LocalMaryInterface.<init>(LocalMaryInterface.java:66)
	at com.jetbrains.idear.tts.TTSService.<init>(TTSService.kt:23)
	at com.jetbrains.idear.tts.TTSService.<clinit>(TTSService.kt:16)
	at com.jetbrains.idear.asr.ASRControlLoop.run(ASRControlLoop.kt:40)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at marytts.util.MaryRuntimeUtils.ensureMaryStarted(MaryRuntimeUtils.java:70)
	at marytts.LocalMaryInterface.<init>(LocalMaryInterface.java:64)
	... 4 more
Exception in thread "ASR Thread" kotlin.KotlinNullPointerException
	at com.jetbrains.idear.tts.TTSService.say(TTSService.kt:46)
	at com.jetbrains.idear.asr.ASRControlLoop.run(ASRControlLoop.kt:40)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -28,6 +28,8 @@ dependencies {
   compile("edu.cmu.sphinx:sphinx4-core:5prealpha-SNAPSHOT")
   compile("com.mashape.unirest:unirest-java:1.4.7")
   compile("org.codehaus.jettison:jettison:1.3.7")
-  compile("de.dfki.mary","voice-cmu-slt-hsmm", "5.2")
+  compile("de.dfki.mary:voice-cmu-slt-hsmm:5.2")
+  compile("com.amazonaws:aws-java-sdk-polly:1.11.77")
+  compile("com.googlecode.soundlibs:jlayer:1.0.1-1")
   testCompile("junit:junit:4.12")
 }

--- a/src/main/java/com/jetbrains/idear/asr/ASRControlLoop.kt
+++ b/src/main/java/com/jetbrains/idear/asr/ASRControlLoop.kt
@@ -5,7 +5,6 @@ import com.intellij.openapi.actionSystem.*
 import com.intellij.openapi.actionSystem.impl.SimpleDataContext
 import com.intellij.openapi.application.ApplicationInfo
 import com.intellij.openapi.application.ApplicationManager
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.util.Pair
 import com.intellij.util.Consumer
 import com.jetbrains.idear.GoogleHelper
@@ -17,7 +16,6 @@ import com.jetbrains.idear.ide.IDEService
 import com.jetbrains.idear.ide.IDEService.invokeAction
 import com.jetbrains.idear.recognizer.CustomLiveSpeechRecognizer
 import com.jetbrains.idear.recognizer.CustomMicrophone
-import com.jetbrains.idear.tts.TTSService
 import com.jetbrains.idear.tts.TTSService.say
 import java.awt.EventQueue
 import java.awt.event.KeyEvent.*

--- a/src/main/java/com/jetbrains/idear/tts/MaryTTS.kt
+++ b/src/main/java/com/jetbrains/idear/tts/MaryTTS.kt
@@ -1,0 +1,69 @@
+package com.jetbrains.idear.tts
+
+import marytts.LocalMaryInterface
+import marytts.MaryInterface
+import marytts.exceptions.MaryConfigurationException
+import marytts.exceptions.SynthesisException
+import marytts.modules.synthesis.Voice
+import marytts.util.data.audio.AudioPlayer
+import java.util.*
+import java.util.logging.Level
+import java.util.logging.Logger
+
+/**
+ * Created by breandan on 7/9/2015.
+ */
+class MaryTTS : TTSProvider {
+    val logger = Logger.getLogger(MaryTTS::class.java.simpleName)
+    private var voice: Voice? = null
+    var maryTTS: MaryInterface? = null
+
+    init {
+        try {
+            maryTTS = LocalMaryInterface()
+            val systemLocale = Locale.getDefault()
+            if (maryTTS!!.availableLocales.contains(systemLocale)) {
+                voice = Voice.getDefaultVoice(systemLocale)
+            }
+
+            if (voice == null) {
+                voice = Voice.getVoice(maryTTS!!.availableVoices.iterator().next())
+            }
+
+            maryTTS!!.locale = voice!!.locale
+            maryTTS!!.voice = voice!!.name
+        } catch (e: MaryConfigurationException) {
+            e.printStackTrace()
+        }
+    }
+
+    override fun say(text: String?) {
+        if (text == null || text.isEmpty()) {
+            return
+        }
+
+        try {
+            val audio = maryTTS!!.generateAudio(text)
+            val player = AudioPlayer(audio)
+            player.start()
+            player.join()
+        } catch (e: SynthesisException) {
+            logger.log(Level.SEVERE, String.format("Sorry! Could not pronounce '%s'", text), e)
+        } catch (e: InterruptedException) {
+            logger.log(Level.SEVERE, String.format("Sorry! Could not pronounce '%s'", text), e)
+        }
+    }
+
+    override fun dispose() {
+    }
+}
+
+fun main(args: Array<String>) {
+    val ttService = MaryTTS()
+    val scan = Scanner(System.`in`)
+
+    while (true) {
+        println("Text to speak:")
+        ttService.say(scan.nextLine())
+    }
+}

--- a/src/main/java/com/jetbrains/idear/tts/PollyTTS.kt
+++ b/src/main/java/com/jetbrains/idear/tts/PollyTTS.kt
@@ -1,0 +1,134 @@
+package com.jetbrains.idear.tts
+
+import com.amazonaws.ClientConfiguration
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain
+import com.amazonaws.regions.Region
+import com.amazonaws.regions.Regions
+import com.amazonaws.services.polly.AmazonPollyClient
+import com.amazonaws.services.polly.model.*
+import java.io.BufferedInputStream
+import java.util.*
+//import javazoom.jl.player.advanced.AdvancedPlayer
+import java.io.IOException
+import java.io.InputStream
+import javax.sound.sampled.*
+import javax.sound.sampled.AudioFormat.Encoding.PCM_SIGNED
+import javax.sound.sampled.SourceDataLine
+import javax.sound.sampled.AudioInputStream
+import javazoom.jl.player.advanced.PlaybackEvent
+import javazoom.jl.player.advanced.PlaybackListener
+import javazoom.jl.player.advanced.AdvancedPlayer
+
+
+
+
+/**
+ * Either:
+ * <ul>
+ *     <li>Users define <code>AWS_ACCESS_KEY_ID</code> (or <code>AWS_ACCESS_KEY</code>) and
+ *          <code>AWS_SECRET_KEY</code> (or <code>AWS_SECRET_ACCESS_KEY</code>)</li>
+ *     <li>the plugin must provide <code>aws.accessKeyId</code> and <code>aws.secretKey</code></li>
+ *     <li>(not implemented) the plugin could use STS or Cognito</li>
+ * </ul>
+ *
+ * @see http://docs.aws.amazon.com/polly/latest/dg/examples-java.html
+ */
+class PollyTTS : TTSProvider {
+    private var polly: AmazonPollyClient? = null
+    private var voice: Voice? = null
+//    private var player: AdvancedPlayer? = null
+
+    init {
+        polly = AmazonPollyClient(DefaultAWSCredentialsProviderChain(), ClientConfiguration())
+        polly!!.setRegion(Region.getRegion(Regions.US_EAST_1))
+
+        // Create describe voices request.
+        val describeVoicesRequest = DescribeVoicesRequest()
+
+        // Synchronously ask Amazon Polly to describe available TTS voices.
+        val describeVoicesResult = polly!!.describeVoices(describeVoicesRequest)
+        voice = describeVoicesResult.voices[0]
+    }
+
+    override fun say(text: String?) {
+        text?.let {
+            val audio = synthesize(it, OutputFormat.Mp3)
+            if (audio != null) {
+                playAudio(audio)
+            }
+        }
+    }
+
+    override fun dispose() {
+    }
+
+    @Throws(IOException::class)
+    private fun synthesize(text: String, format: OutputFormat): InputStream? {
+        val synthReq = SynthesizeSpeechRequest().withText(text).withVoiceId(voice?.getId())
+                .withOutputFormat(format)
+        val synthRes = polly?.synthesizeSpeech(synthReq)
+
+        return synthRes?.audioStream
+    }
+
+    private fun playAudio(inputStream: InputStream) {
+        //create an MP3 player
+        val player = AdvancedPlayer(inputStream,
+                javazoom.jl.player.FactoryRegistry.systemRegistry().createAudioDevice())
+
+        player.playBackListener = object : PlaybackListener() {
+            override fun playbackStarted(evt: PlaybackEvent?) {
+                println("Playback started")
+            }
+
+            override fun playbackFinished(evt: PlaybackEvent?) {
+                println("Playback finished")
+            }
+        }
+
+        // play it!
+        player.play()
+
+        /* Borrowed from http://blog.conygre.com/2016/12/06/at-the-third-stroke-the-time-will-be-spoken-by-aws-polly/
+        ...but it didn't work
+
+        val audio = AudioSystem.getAudioInputStream(BufferedInputStream(inputStream))
+        val format = getOutFormat(audio.format)
+        val info = DataLine.Info(SourceDataLine::class.java, format)
+
+        val line = AudioSystem.getLine(info) as SourceDataLine
+        if (line != null) {
+            line!!.open(format)
+            line!!.start()
+            stream(AudioSystem.getAudioInputStream(format, audio), line)
+            line!!.drain()
+            line!!.stop()
+        }*/
+    }
+
+    /*private fun getOutFormat(inFormat: AudioFormat): AudioFormat {
+        val ch = inFormat.channels
+        val rate = inFormat.sampleRate
+        return AudioFormat(PCM_SIGNED, rate, 16, ch, ch * 2, rate, false)
+    }
+
+    @Throws(IOException::class)
+    private fun stream(`in`: AudioInputStream, line: SourceDataLine) {
+        val buffer = ByteArray(65536)
+        var n = 0
+        while (n != -1) {
+            line.write(buffer, 0, n)
+            n = `in`.read(buffer, 0, buffer.size)
+        }
+    }*/
+}
+
+fun main(args: Array<String>) {
+    val ttService = PollyTTS()
+    val scan = Scanner(System.`in`)
+
+    while (true) {
+        println("Text to speak:")
+        ttService.say(scan.nextLine())
+    }
+}

--- a/src/main/java/com/jetbrains/idear/tts/TTSProvider.kt
+++ b/src/main/java/com/jetbrains/idear/tts/TTSProvider.kt
@@ -1,0 +1,6 @@
+package com.jetbrains.idear.tts
+
+interface TTSProvider {
+    fun say(text: String?)
+    fun dispose()
+}

--- a/src/main/java/com/jetbrains/idear/tts/TTSService.kt
+++ b/src/main/java/com/jetbrains/idear/tts/TTSService.kt
@@ -1,71 +1,22 @@
 package com.jetbrains.idear.tts
 
-import marytts.LocalMaryInterface
-import marytts.MaryInterface
-import marytts.exceptions.MaryConfigurationException
-import marytts.exceptions.SynthesisException
-import marytts.modules.synthesis.Voice
-import marytts.util.data.audio.AudioPlayer
-import java.util.*
-import java.util.logging.Level
-import java.util.logging.Logger
-
-/**
- * Created by breandan on 7/9/2015.
- */
 object TTSService {
-    val logger = Logger.getLogger(TTSService::class.java.simpleName)
-    private var voice: Voice? = null
-    var maryTTS: MaryInterface? = null
+    var ttsProvider: TTSProvider? = null
 
     init {
-        try {
-            maryTTS = LocalMaryInterface()
-            val systemLocale = Locale.getDefault()
-            if (maryTTS!!.availableLocales.contains(systemLocale)) {
-                voice = Voice.getDefaultVoice(systemLocale)
-            }
-
-            if (voice == null) {
-                voice = Voice.getVoice(maryTTS!!.availableVoices.iterator().next())
-            }
-
-            maryTTS!!.locale = voice!!.locale
-            maryTTS!!.voice = voice!!.name
-        } catch (e: MaryConfigurationException) {
-            e.printStackTrace()
-        }
+//        ttsProvider = MaryTTS()
+        ttsProvider = PollyTTS()
     }
 
-    fun say(text: String?) {
-        if (text == null || text.isEmpty()) {
-            return
-        }
+    // TODO: list voices by locale
+    // TODO: allow user to select voice
 
-        try {
-            val audio = maryTTS!!.generateAudio(text)
-            val player = AudioPlayer(audio)
-            player.start()
-            player.join()
-        } catch (e: SynthesisException) {
-            logger.log(Level.SEVERE, String.format("Sorry! Could not pronounce '%s'", text), e)
-        } catch (e: InterruptedException) {
-            logger.log(Level.SEVERE, String.format("Sorry! Could not pronounce '%s'", text), e)
-        }
+    fun say(text: String?) {
+        ttsProvider?.say(text)
     }
 
     fun dispose() {
-
-    }
-
-}
-
-fun main(args: Array<String>) {
-    val ttService = TTSService
-    val scan = Scanner(System.`in`)
-
-    while (true) {
-        println("Text to speak:")
-        ttService.say(scan.nextLine())
+        ttsProvider?.dispose()
+        ttsProvider = null
     }
 }

